dra: kubelet must skip NodePrepareResource if not used by any container

pohly opened the pull request:
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

If (for whatever reason) no container uses a claim, then there's no need to prepare it. We also need to test this.

#### Does this PR introduce a user-facing change?
```release-note
As in Kubernetes 1.26 and 1.27, resource claims do not get prepared by kubelet when no container uses them. This was changed accidentally in [v1.28.0-alpha.1](https://github.com/kubernetes/kubernetes/releases/tag/v1.28.0-alpha.1).
```

pohly committed:
e2e dra: add "kubelet must skip NodePrepareResource if not used by any container"

If (for whatever reason) no container uses a claim, then there's no need to
prepare it.

pohly commented:
This is a draft because it fails for me locally. kubelet calls `NodePrepareResource`.

/assign @bart0sh

pohly committed:
kubelet dra: restore skipping of unused resource claims

1aeec10efb removed iterating over containers in favor of iterating over pod
claims. This had the unintended consequence that NodePrepareResource gets
called unnecessarily when no container needs the claim. The more natural
behavior is to skip unused resources. This enables (theoretic, at this time)
use cases where some DRA driver relies on the controller part to influence
scheduling, but then doesn't use CDI with containers.

pohly commented:
/kind bug

pohly commented:
/unassign @bart0sh 
/assign @klueska

pohly commented:
/retest

klueska commented on pkg/kubelet/cm/dra/manager.go:
```diff
@@ -99,6 +100,13 @@ func (m *ManagerImpl) PrepareResources(pod *v1.Pod) error {
 				pod.Name, pod.UID, claimName, resourceClaim.UID)
 		}
 
+		// If no container actually uses the claim, then we don't need
+		// to prepare it.
+		if !claimIsUsedByPod(podClaim, pod) {
+			klog.V(5).InfoS("Skipping unused resource", "claim", claimName, "pod", pod.Name)
+			continue
+		}
+
```
Something seems off about the placement of this check in this function.

I _believe_ it still works, but it may unnecessarily delay the underyling `NodeUnprepareResources()` in certain scenarios.

The scenarios to consider are:
1. This is the only pod to reference the claim
2. This is the first pod to reference the claim (and more than one eventually reference it)
3. This is not the first pod to reference the claim

In scenario 1, `NodePrepareResources()` will never be called (because of the new check), and `NodeUnprepareResources()` will also never be called (because a claimInfo object for it will never be created).

In scenario 2, `NodePrepareResources()` will not be called when the first pod is started (because of the new check), a reference to the pod will also *not* be added to the `claimInfo` object. Future calls to `PrepareResources()` will have their pods added to the `claimInfo` as appropriate and future calls to `UnprepareResources()` will only trigger an underlying `NodeUnprepareResources()` once the last pod reference in the claimInfo has called it.

In scenario 3, `NodePrepareResources()` will have already been trigged by a previous pod, and since our new check for `!claimIsUsedByPod()` is further down in the code, a reference to this pod *will* be added to the `claimInfo` object. Future calls to `UnprepareResources()` will only trigger an underlying `NodeUnprepareResources()` once the last pod reference in the claimInfo has called it (even if the only pod we are waiting for is one that would fail the `!claimIsUsedByPod()` check.

pohly replied:
I placed it here because I wanted to do the "is reserved for" check also when the claim is not actually used. We still want to do proper scheduling despite that, which means the claim must be reserved for the pod to run.

What I missed is the 
```
			claimInfo.addPodReference(pod.UID)
			continue
```
above when the claim is already prepared.

I think that's actually another bug in the code: if the claim was prepared for some other pod and then this pod here does't have it reserved, it's allowed to run by kubelet because the "is reserved for" check gets skipped.

pohly replied:
I fixed this by moving the "is already prepared" check.

klueska replied:
Seems reasonable. I believe the initial thinking was to exit early to avoid a call to the API server with `ResourceV1alpha2().ResourceClaims(pod.Namespace).Get()`, but I agree that checking `reservedFor` is more important than this minor optimization.

pohly committed:
kubelet dra: fix checking of second pod which uses a claim

When a second pod wanted to use a claim, the obligatory sanity check whether
the pod is really allowed to use the claim ("reserved for") was skipped.
